packages/react-reconciler/src/__tests__/ReactSuspenseFallback-test.js

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+let React;
+let ReactTestRenderer;
+let ReactFeatureFlags;
+let Scheduler;
+let ReactCache;
+let Suspense;
+
+let TextResource;
+let textResourceShouldFail;
+
+describe('ReactSuspenseFallback', () => {
+  beforeEach(() => {
+    jest.resetModules();
+    ReactFeatureFlags = require('shared/ReactFeatureFlags');
+
+    ReactFeatureFlags.replayFailedUnitOfWorkWithInvokeGuardedCallback = false;
+    React = require('react');
+    ReactTestRenderer = require('react-test-renderer');
+    Scheduler = require('scheduler');
+    ReactCache = require('react-cache');
```

Comment:
Can you use the helpers in ReactSuspenseWithNoopRenderer-test or ReactCache-test instead? Those are the  closest to our preferred testing idioms. We're going to remove the `react-cache` package.

Reply:
Ideally just put them into those existing files. It tends to get annoying to update the helpers in different files that use the same patterns. For example we'll need to do a bunch of refactoring to change the avoidThisFallback API that needs to basically test all the same things again.

Reply:
I put them in separate files because @acdlite mentioned that we should break out more tests into separate files for speed and organizational reasons.

Reply:
In that case, maybe the helpers should be broken out into a shared module too.

Organizationally I find that what organization we thought would make sense in terms of grouping features doesn't hold up for the next change.

The defacto pattern that has evolved that is pretty convenient is that files are grouped by the kind of test mechanism. So you can find a test with similar mechanism that you need and then copy it. It doesn't help speed so the way we solved that for the ServerIntegration tests is that we broke them out into a shared helper and then pulled that in everywhere.

It would also be nice if we didn't mix test mechanisms in the same file.

Reply:
My main reason for suggesting more focused module was because some of them are getting really slow.

I've been resistant to the idea of having test helper modules because I don't want it to become a dumping ground for a bunch of idiosyncratic abstractions. I like that if there's a bad abstraction, it's confined to a single module.

Reply:
Really it's because I dislike when a test breaks and I have to open multiple files to figure out what the hell is going on, which is the case for some of our older server rendering tests

Reply:
The scheduler.yield things are a helper that is an example of doing it right so it can be done right.

I'm not a fan of the server rendering strategy per se but at least it's the same one every time. The ones that get really frustrating to upgrade from an older pattern are the ones using similar but slightly different patterns.

Who am I and what have I done to the good olde wet Seb?

Reply:
Ultimately this is kind of a flaw of jest. We reset modules between every run anyway. There's no reason we couldn't run every individual test case in a separate process. That would also let us use proper ES modules since it would by default reset modules.

Reply:
> The ones that get really frustrating to upgrade from an older pattern are the ones using similar but slightly different patterns.

I agree with that but I suppose I don't mind because in aggregate we spend more time debugging tests than we do making API changes.

In this particular case, though, I think the "async text" pattern has solidified enough that I'd be comfortable moving it to its own helper, a la `yieldValue`.